vsphere/internal/virtualdevice/virtual_machine_device_subresource.go

```diff
@@ -617,6 +659,24 @@ func setSCSIBusSharing(l *object.VirtualDeviceList, ctlr types.BaseVirtualSCSICo
 	return cspec, nil
 }
 
+// createIDEController creates a new IDE controller.
+func createIDEController(l *object.VirtualDeviceList, bus int) ([]types.BaseVirtualDeviceConfigSpec, error) {
+	ide, _ := l.CreateIDEController()
+	cspec, err := object.VirtualDeviceList{ide}.ConfigSpec(types.VirtualDeviceConfigSpecOperationAdd)
+	*l = applyDeviceChange(*l, cspec)
```

Comment:
Idle thought: how much does this API request "cost" in terms of time ? Is there a way to gather all controller changes and apply them in a single go ?

Reply:
The changes being applied at this point are actually all to a virtual set of devices, so there aren't any API calls to do these. After all the device changes are processed, a reconfig spec is generated and then finally applied to the real VM on vSphere here: https://github.com/hashicorp/terraform-provider-vsphere/blob/master/vsphere/resource_vsphere_virtual_machine.go#L748.